tutorials/pytest.md

```diff
@@ -0,0 +1,191 @@
+
+Zauważmy, że pojawiło się wyrażenie [assert](https://www.w3schools.com/python/ref_keyword_assert.asp). W skrócie, jest to warunek `if` zarezerwowany do testowania, tj.
+```
+x = "hello"
+
+#if condition returns True, then nothing happens:
+assert x == "hello"
+
+#if condition returns False, AssertionError is raised:
+assert x == "goodbye"
+```
+
+Wracając do naszego testu, napisaliśmy test sprawdzający, czy dla prawidłowych argumentów, otrzymujemy wynik większy od 0. Dodatkowo, aby móc to osiągnąć, musieliśmy sięgnąć do folderu z aplikacją, żeby wyciągnąć testowaną klasę, tj. `from app.bmi_calc import Osoba`. Przetestujmy!
+
+![img](pics/pytest5.png)]]
+
+Sukces!
+
+Mozemy zabrać się za pisanie drugiego testu. Teraz spróbujemy napisać test jednostkowy, który sprawdzałby, czy jest napisane ograniczenie (zwracające błąd) w przypadku wprowadzenia nieprawdiłowych wartości, tj. mniejszych od 0.
+```
+import pytest
+from app.bmi_calc import Osoba
+
+def test_first():
+    pass
+
+def test_for_positive_bmi():
+    bmi = Osoba(1, 187, 82).BMI()
+    assert bmi > 0
+
+def test_for_positive_bmi_values_error_raise():
+
+    with pytest.raises(ValueError):
+        bmi = Osoba(1, 187, -82).BMI()
+```
+
+Zwróć uwagę, że tym razem zaimportowaliśmy bilbiotekę `pytest`. Dzieje się tak ponieważ teraz chcemy przetestować konkretną funkcjonalność tej bilbioteki. Test, który napisaliśmy, sprawdza, czy w przypadku wprowadznie nieprawidłowych wartości `bmi = Osoba(1, 187, -82).BMI()`, aplikacja zwróci nam błąd typou `ValueError`, tj. `with pytest.raises(ValueError)`.
+
+Przetestujmy!
+
+![img](pics/pytest6.png)]]
+
+Porażka!
+
+Rzucając okiem na naszą aplikację, faktycznie brakuje tam wyjątku, który reagowałby błędem w przypadku wprowadznie nieprawidłowych wartości. Naprawmy to.
+```
+import random
+
+class Osoba:
+    def __init__(self, plec: bin, wzrost: int, waga: int): # 1 - male, 0 - female
+        self.wzrost = wzrost
+        self.waga = waga
+        self.plec = plec
+    
+    def BMI(self):
+        BMI = self.waga/(self.wzrost/100)**2
+
+        if BMI <= 0:
+            raise ValueError("BMI jest ujemne!")
+
+        if self.plec != 1:
+            return BMI*0.95
+        else:
+            return BMI
+```
+
+Przetestujmy jeszcze raz naszą aplikację.
+
+![img](pics/pytest7.png)]]
+
+![img](pics/Great_Success.jpg)]]
+
+*(W przypadkach prawdziwych projektów, euforia związana ze "wszystkimi testami na zielono" jest znacznie większa niż ta przedstawiona na obrazku.)*
+
+Oficjalnie jesteśmy testerami! Udało nam się napsać bardzo prostą aplikację oraz zabezpieczyć ją o dwa podstawowe testy jednostkowe. Zwróć również uwagę, że w przypadku drugiego testu, postąpiliśmy *zgodnie ze sztuką* pisania testów jednostkowych.
+
+> Jeżeli nie jesteśmy zwolennikami terminala, możemy wywoływać testy z poziomu notatnika uruchomiwszy funkcję `!pytest ../tests` w komórce. Będzie to odpowiadało uruchamianiu komendy `pytest` z poziomu folderu głównego, z użyciem terminala.
```

Comment:
próbowałeś uruchamiać pytest w visual studio code?

Reply:
Tak, działają, ale to wymaga:
1. Wtyczki w VS code.
2. Większej wprawy w testowaniu.

Reply:
![Zrzut ekranu 2022-11-19 145426](https://user-images.githubusercontent.com/51819246/202854082-e5f432f3-8e3c-4887-9b10-57d0fcc4ef21.png)